pyqir-generator/src/python.rs

```diff
@@ -1,174 +1,264 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
-use crate::emit::{get_bitcode_base64_string, get_ir_string, write_model_to_file};
-use crate::interop::{
-    ClassicalRegister, Controlled, Instruction, Measured, QuantumRegister, Rotated, SemanticModel,
-    Single,
+
+use crate::{
+    emit,
+    interop::{
+        ClassicalRegister, Controlled, If, Instruction, Measured, QuantumRegister, Rotated,
+        SemanticModel, Single,
+    },
 };
-use log;
-use pyo3::exceptions::PyOSError;
-use pyo3::prelude::*;
+use pyo3::{exceptions::PyOSError, prelude::*};
 
 #[pymodule]
-fn pyqir_generator(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
-    m.add_class::<PyQIR>()?;
-    Ok(())
+fn _native(_py: Python, m: &PyModule) -> PyResult<()> {
+    m.add_class::<SimpleModule>()?;
+    m.add_class::<Qubit>()?;
+    m.add_class::<Ref>()?;
+    m.add_class::<Builder>()?;
+    m.add_class::<BasicQisBuilder>()
 }
 
+const RESULT_NAME: &str = "result";
+const QUBIT_NAME: &str = "qubit";
+
 #[pyclass]
-pub struct PyQIR {
-    pub(super) model: SemanticModel,
+struct SimpleModule {
+    model: SemanticModel,
+    builder: Py<Builder>,
 }
 
 #[pymethods]
-impl PyQIR {
+impl SimpleModule {
     #[new]
-    fn new(name: String) -> Self {
-        PyQIR {
-            model: SemanticModel::new(name),
+    fn new(py: Python, name: String, num_qubits: u64, num_results: u64) -> PyResult<SimpleModule> {
+        let registers = vec![ClassicalRegister::new(RESULT_NAME.to_string(), num_results)];
+
+        let qubits = (0..num_qubits)
+            .map(|i| QuantumRegister::new(QUBIT_NAME.to_string(), i))
+            .collect();
```

Comment:
Out of curiosity, it looks like these two definitions are fairly asymmetrical — is there a reason `registers` is defined as containing a single 𝑛-bit classical register while `qubits` contains 𝑛 1-qubit registers?

Reply:
@idavis defined the register types and would know more.

I agree with you that QuantumRegister and ClassicalRegister should be symmetric. There is also the issue of the instructions in the interop module using string identifiers, which requires the new python module to unnecessarily allocate strings. I would like to clean up these issues in a separate PR.

Reply:
#53